writer_test.go

```diff
@@ -11,16 +24,72 @@ import (
 func TestRecordMetadata(t *testing.T) {
 	is := is.New(t)
 
-	sdkMetadata := sdk.Metadata{
+	var rec sdk.Record
+	rec.Metadata = sdk.Metadata{
 		"created_at":     "2023-03-15T14:25:07Z",
 		"pinecone.prop1": "a1",
 		"pinecone.prop2": "a2",
 	}
-	recMetadata, err := recordMetadata(sdkMetadata)
+	recMetadata, err := parsePineconeMetadata(rec)
 	is.NoErr(err)
 
-	fmt.Println()
-
 	is.Equal(recMetadata.Fields["prop1"].AsInterface().(string), "a1")
 	is.Equal(recMetadata.Fields["prop2"].AsInterface().(string), "a2")
 }
+
+func testRecord(op sdk.Operation) sdk.Record {
+	var position sdk.Position
+	var key sdk.Data = sdk.RawData("key")
+	var metadata sdk.Metadata
+	var payload sdk.Data = sdk.StructuredData{
+		"vector": []float64{1, 2},
+	}
+
+	return sdk.Record{
+		Position: position, Operation: op,
+		Metadata: metadata, Key: key,
+		Payload: sdk.Change{
+			Before: nil,
+			After:  payload,
+		},
+	}
+}
+
+func batchSize(batch recordBatch) int {
+	switch batch := batch.(type) {
+	case upsertBatch:
+		return len(batch.vectors)
+	case deleteBatch:
+		return len(batch.ids)
+	}
+
+	panic("invalid batch")
+}
+
+func TestParseRecords(t *testing.T) {
+	is := is.New(t)
+
+	records := []sdk.Record{
+		testRecord(sdk.OperationUpdate),
+
+		testRecord(sdk.OperationDelete), testRecord(sdk.OperationDelete),
+
+		testRecord(sdk.OperationCreate), testRecord(sdk.OperationCreate),
+		testRecord(sdk.OperationCreate),
+
+		testRecord(sdk.OperationDelete),
+
+		testRecord(sdk.OperationSnapshot), testRecord(sdk.OperationSnapshot),
+	}
+
+	batches, err := buildBatches(records)
+	is.NoErr(err)
+
+	is.Equal(len(batches), 5)
+
+	is.Equal(batchSize(batches[0]), 1)
+	is.Equal(batchSize(batches[1]), 2)
+	is.Equal(batchSize(batches[2]), 3)
+	is.Equal(batchSize(batches[3]), 1)
+	is.Equal(batchSize(batches[4]), 2)
```

Comment:
Having the batch sizes checked helps a lot. :+1:  Here it would be good to also check if the batches themselves are correct. For example, the fact that we have 2 vectors in a delete batch doesn't mean the IDs are correct (and we might delete wrong IDs).

Reply:
Yeah, agree, it is a bit weak. I just added new tests.